src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala

```diff
@@ -438,27 +414,19 @@ class SchedulerActions(
     instancesToKill.foreach { instances: Seq[Instance] =>
       logger.info(s"Scaling ${runSpec.id} from ${runningInstances.size} down to $targetCount instances")
 
-      launchQueue.purge(runSpec.id)
-        .recover {
-          case NonFatal(e) =>
-            logger.warn("Async purge failed: {}", e)
-            Done
-        }.foreach { _ =>
-          logger.info(s"Killing instances ${instances.map(_.instanceId)}")
-
-          def killInstances(instances: Seq[Instance]): Future[Done] = async {
-            val changeGoalsFuture = instances.map { i =>
-              if (i.hasReservation) instanceTracker.setGoal(i.instanceId, Goal.Stopped)
-              else instanceTracker.setGoal(i.instanceId, Goal.Decommissioned)
-            }
-
-            await(Future.sequence(changeGoalsFuture))
-            await(killService.killInstances(instances, KillReason.OverCapacity))
-          }
-
-          killInstances(instances)
-        }
+      val instancesAreTerminal = killService.watchForKilledInstances(instances)
```

Comment:
Note for the future: Recently, we introduced a subscription mechanism for InstanceTracker. It'd be cool to update the watcher to use this as there is much less possibility for a race (subscription mechanism gives a consistent view of InstanceTracker state, regardless if the subscription is registered before or after the terminal status update is received).

Reply:
why can't we iterate on this? So far we were doing it this way, I don't see why we would have to change it in this PR. I am totally open for puting that improvement into JIRA and working on it right after we merge this one...

Reply:
> Note for the future

As in... not now. As in, yes, iterate.

Reply:
(I agree with Alena; the behavior stays more-or-less the same by using the watchForKilledInstances method.)

Reply:
> The watches must be registered before the kill is sent, otherwise you could miss the terminal task status update.

I get that but why not check for the instances to be terminal?

Reply:
Sorry I don't get your comment. Also I don't really see why would we want to change that in this PR...

Reply:
Instead of waiting for an instance update we could simply query the instance tracker and check the state of the instances.